bee-node/bee-plugin/src/hotloader.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Hotloading related utilities.
+
+use crate::{PluginError, PluginId, PluginManager, UniqueId};
+
+use bee_event_bus::EventBus;
+use tokio::{
+    process::Command,
+    time::{sleep, Duration},
+};
+
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    sync::Arc,
+    time::SystemTime,
+};
+
+const PLUGIN_CHECK_INTERVAL_MILLIS: u64 = 1000;
+
+/// Type that watches a directory for changes and loads every file as a plugin.
+pub struct PluginHotloader {
+    directory: PathBuf,
+    plugins_info: HashMap<PathBuf, PluginInfo>,
+    manager: PluginManager,
+}
+
+impl PluginHotloader {
+    /// Creates a new [`PluginHotloader`] that watches the specified directory.
+    pub fn new<P: AsRef<Path> + ?Sized>(directory: &P, event_bus: Arc<EventBus<'static, UniqueId>>) -> Self {
+        Self {
+            directory: directory.as_ref().to_owned(),
+            plugins_info: HashMap::new(),
+            manager: PluginManager::new(event_bus),
+        }
+    }
+
+    /// Watches for changes in a directory and loads new plugins.
+    ///
+    /// This is done by following these rules:
+    /// - If a file is created, it will be loaded and executed as a plugin.
+    /// - If a file is removed, the process for it will be terminated and the plugin will be unloaded.
+    /// - If a file is modified, it will behave as if the file was removed and created in succession.
+    pub async fn run(mut self) -> Result<(), PluginError> {
+        loop {
+            let mut dir = tokio::fs::read_dir(&self.directory).await?;
+
+            let mut last_writes = HashMap::new();
+
+            while let Some(entry) = dir.next_entry().await? {
+                let path = entry.path();
+                let metadata = entry.metadata().await?;
+
+                if metadata.is_file() {
```

Comment:
This doesn't seem sufficient to me, it will try to execute regular files and crash

Reply:
yeah but it will be a child process crashing so we will get a warning in the logs and that's it.